components/org.wso2.carbon.identity.conditional.auth.functions.http/src/main/java/org/wso2/carbon/identity/conditional/auth/functions/http/util/ClientCredentialAuthConfig.java

```diff
@@ -0,0 +1,320 @@
+        return tokenEndpoint;
+    }
+
+    public String getConsumerKey() {
+        return consumerKey;
+    }
+
+    public String getConsumerSecret() {
+        return consumerSecret;
+    }
+
+    public String getScopes() {
+        return scopes;
+    }
+
+    @Override
+    public HttpUriRequest applyAuth(HttpUriRequest request, AuthConfigModel authConfigModel)
+            throws FrameworkException, SecretManagementException {
+
+        this.apiAccessTokenCache = APIAccessTokenCache.getInstance();
+        Map<String, Object> properties = authConfigModel.getProperties();
+        validateRequiredProperties(properties);
+
+        setConsumerKey(properties.get(CONSUMER_KEY_VARIABLE_NAME).toString());
+        setConsumerSecret(properties.get(CONSUMER_SECRET_VARIABLE_NAME).toString());
+        setTokenEndpoint(properties.get(TOKEN_ENDPOINT).toString());
+        setScopes(properties.containsKey(SCOPES) ? properties.get(SCOPES).toString() : null);
+
+        String accessToken = getAccessToken();
+        if (accessToken == null) {
+            asyncReturn.accept(authenticationContext, Collections.emptyMap(), OUTCOME_FAIL);
+            LOG.error("Failed to retrieve access token. Aborting request.");
+            throw new FrameworkException("Failed to retrieve access token.");
+        }
+        request.setHeader(AUTHORIZATION, BEARER + accessToken);
+        return request;
+    }
+
+    /**
+     * This method is used to get the access token from the token endpoint.
+     *
+     * @return Access token
+     * @throws FrameworkException {@link FrameworkException}
+     */
+    private void validateRequiredProperties(Map<String, Object> properties) throws FrameworkException {
+
+        if (!properties.containsKey(CONSUMER_KEY_VARIABLE_NAME) ||
+                !properties.containsKey(CONSUMER_SECRET_VARIABLE_NAME) ||
+                !properties.containsKey(TOKEN_ENDPOINT)) {
+            asyncReturn.accept(authenticationContext, Collections.emptyMap(), OUTCOME_FAIL);
+            LOG.error("Required properties not defined. Aborting token request.");
+            throw new FrameworkException("Missing required properties.");
+        }
+    }
+
+    /**
+     * This method is used to get the access token from the cache or request a new token from the token endpoint.
+     *
+     * @return Access token
+     * @throws FrameworkException {@link FrameworkException}
+     */
+    private String getAccessToken() throws FrameworkException {
+        String accessToken = apiAccessTokenCache.getValueFromCache(getConsumerKey(),
+                authenticationContext.getTenantDomain());
+        try {
+            if (StringUtils.isNotEmpty(accessToken) && !isTokenExpired(accessToken)) {
+                LOG.debug("Unexpired access token available in cache.");
+                return accessToken;
+            } else {
+                return attemptAccessTokenRequest(MAX_TOKEN_REQUEST_ATTEMPTS_FOR_TIMEOUT);
```

Comment:
Use same retry count as in external call

Reply:
Fixed with https://github.com/wso2-extensions/identity-conditional-auth-functions/pull/157/commits/afe4342c547dfe83345abfd23f7fee114389532b